contribute/design/openebs-provisioner-high-availibility.md

```diff
@@ -0,0 +1,98 @@
+# High Availability for OpenEBS Volume Provisioner
+---
+## Problems addressed:
+
+- `openebs-provisioner` should always be available.
+- `openebs-provisioner` should always be running on the specified/labeled pods. 
```

Comment:
openebs-provisioner is itself a pod, what do you mean by running on specified/labeled pods?

Reply:
@satyamz -- can you clarify the above question?

Reply:
@kmova It should be:  Each node should always be running the specified/labeled `openebs-provisioner`  pods on each node provided that replica count is equal to the number of nodes.

Reply:
@satyamz -- As per the yaml spec - kubernetes scheduler will try to run 2 pods of openebs-provisioner when it can. 

What is the behavior you expect in the following cases w.r.t to scheduler:
case1: there is only 1 node in the kubernetes cluster - with resources available to run openebs-provisioner
case2: there are 2 nodes in the kubernetes cluster - with resources available to run openebs-provisioner
case3:- there are 3 nodes in the kubernetes cluster - with resources available to run openebs-provisioner

Also, in case1, will it try to evacuate or stop --other low priority pods to make room for openebs-provisioner? 

In case the scheduler is unable to schedule openebs-provisioner, how will the user know what actions to be taken to rectify the situation.

Reply:
This PR will make sure that at least one `openebs-provisioner` pod is running on each node (considering `replicas` == the number of nodes ). We are trying to solve the problem where `openebs-provisioner` pods are running on all nodes. Hence, pod anti-affinity:
> Pod anti-affinity can prevent the scheduler from locating a new pod on the same node as pods with the same labels if the label selector on the new pod matches the label on the current pod. 

Regarding above cases:
>What is the behavior you expect in the following cases w.r.t to scheduler:
case1: there is only 1 node in the kubernetes cluster - with resources available to run openebs-provisioner
case2: there are 2 nodes in the kubernetes cluster - with resources available to run openebs-provisioner
case3:- there are 3 nodes in the kubernetes cluster - with resources available to run openebs-provisioner
Also, in case1, will it try to evacuate or stop --other low priority pods to make room for openebs-provisioner?

For this, we need to have PDB & PriorityClass for `openebs-provisioner` and other OpenEBS components. I will create a new issue for these. (This was also suggested by Amit)

Reply:
@kmova  created issue https://github.com/openebs/openebs/issues/737 for this and these cases will be covered in different PR.